### DISCUSSION TITLE
Cleaning up failed job containers

### DISCUSSION THREAD

**Discussion by drcrees** [concourse/concourse#6509] [Help & Support] [answered]
We currently have a few jobs that hold 200+ tasks that run simple flyway migrations (hub and spoke) and are running into an upper container limit on our Linux workers, ~200 or so.  

I've read that we can increase the number of containers a Linux worker can run at a time, but that is not advised since the default is around what Linux itself can support.

This is really only a problem when these jobs fail and keep _all_ the containers around for intercept/debugging purposes.
I'm wondering if there is a way we can tell Concourse to **not** keep the containers running and terminate them immediately after job completion, success or fail.

**Comment by taylorsilva** [comment]
I thought there was no way to change this behaviour BUT I did find this flag in the web that alludes to that failed containers do get gc'd:
```
--gc-failed-grace-period=        Period after which failed containers will be garbage collected (default: 120h) [$CONCOURSE_GC_FAILED_GRACE_PERIOD]
```

You could bring this down to maybe `5m` which is the same as hijack grace period.

**Reply by taylorsilva** [reply]
Locally adjusted that setting and failed containers (red builds) did disappear after the grace period passed!

**Reply by drcrees** [reply]
Thanks @taylorsilva!  That is something to consider, though it is not as granular as we would like.
It looks like this will affect _all_ jobs.  It is still nice to have containers stick around for some jobs, but not others.

**Comment by chenbh** [comment, marked as answer]
> I've read that we can increase the number of containers a Linux worker can run at a time, but that is not advised since the default is around what Linux itself can support.

Concourse defaults to a max of 250 containers, this is more of a leftover from Garden/CloudFoundry days rather than any OS constraints. It should be safe to bump it up depending on how much resources your system has, the folks over at EngineerBetter did a great article on it a while back: https://www.engineerbetter.com/blog/increasing-concourse-container-limits/

**Reply by drcrees** [reply]
Thanks @chenbh!  I have gone through that article
My concern with that approach stems from this comment by @jtarchie

https://github.com/concourse/concourse/issues/1620#issuecomment-330551451

> I would recommend that you don't increase your worker to any more that ~250 containers. Linux just starts having a bad time handling and scheduling that many containers. You might have reached the point that adding another worker would best suit your needs. 

Adding workers would definitely help, but then it turns into a cloud cost problem.

**Reply by jtarchie** [reply]
My knowledge is purely based on googling and [stackoverflow](https://stackoverflow.com/a/37628520). There’s no _exact_ number max containers, just contention.
